### PR TITLE
NIFI-10775 Improve support for DescribedValue in PropertyDescriptor.Builder

### PR DESCRIPTION
**Pull request by nandorsoma**
# Summary

[NIFI-10775](https://issues.apache.org/jira/browse/NIFI-10775)

This PR adds support for DescribedValue in the `dependsOn()` and in the `defaultValue()` builder methods. Unfortunately, the new signature `defaultValue(null)`  doesn’t work. Only if you cast it. Because of that, I’ve removed the redundant null defaultValue assignments. However, at this point, I’m unsure what brings more value, and I cannot decide it at the moment. Sometimes it is useful to indicate that the default value is null. On the other hand, reducing redundancy and getting rid of unnecessary getValues() contribute to clean code. A workaround could be to add `defaultNull()` to the Builder() that can be used instead of `defaultValue(null)`. That could be useful in the case of `sensitive(boolean)` too.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Review by turcsanyip — CHANGES_REQUESTED** [review]
Thanks for this improvement @nandorsoma!
Overall it looks good to me.
Found only minor issues in the javadoc and the test.

File: nifi-api/src/main/java/org/apache/nifi/components/PropertyDescriptor.java [hunk 1]
```diff
@@ -591,6 +606,38 @@ public Builder dependsOn(final PropertyDescriptor property, final String firstDe
...
 
+        /**
+         * Establishes a relationship between this Property and the given property by declaring that this Property is only relevant if the given Property has a value equal to one of the given
+         * {@link DescribedValue} arguments.
+         * If this method is called multiple times, each with a different dependency, then a relationship is established such that this Property is relevant only if all dependencies are satisfied.
+         *
+         * In the case that this property is NOT considered to be relevant (meaning that it depends on a property whose value is not specified, or whose value does not match one of the given
+         * Described Values), the property will not be shown in the component's configuration in the User Interface. Additionally, this property's value will not be considered for
+         * validation. That is, if this property is configured with an invalid value and this property depends on Property Foo, and Property Foo does not have a value set, then the component
+         * will still be valid, because the value of this property is irrelevant.
```
[the remark this answers is not in the document]
Comment (turcsanyip) [reply nifi-api/src/main/java/org/apache/nifi/components/PropertyDescriptor.java:617]: I misunderstood the javadoc, no change needed.
Please ignore my comment.

**Comment by nandorsoma** [comment]
Thanks for the review @turcsanyip! I've addressed your comments, please see my latest commit.

File: nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/notification/http/HttpNotificationService.java [hunk 2]
```diff
@@ -127,14 +124,6 @@ public class HttpNotificationService extends AbstractNotificationService {
             .sensitive(true)
             .build();
 
-    public static final PropertyDescriptor SSL_ALGORITHM = new PropertyDescriptor.Builder()
```
Comment (Lehel44) [inline nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/notification/http/HttpNotificationService.java:130]: Was this intentional?

File: nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/notification/http/HttpNotificationService.java [hunk 3]
```diff
@@ -92,15 +92,13 @@ public class HttpNotificationService extends AbstractNotificationService {
     public static final PropertyDescriptor PROP_TRUSTSTORE_PASSWORD = new PropertyDescriptor.Builder()
             .name("Truststore Password")
             .description("The password for the Truststore")
-            .defaultValue(null)
```
Comment (Lehel44) [inline nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/notification/http/HttpNotificationService.java:95]: Does removing these null default values come from any of the current changes?

**Review by Lehel44 — CHANGES_REQUESTED** [review]
Thanks @nandorsoma for the contribution. I provided a few minor comments about the Javadoc. Functionally looks good. 

Since I noticed at the end that "this property", "given property" and "dependent" are used incorrectly or ambiguously throughout the class, consider my comments on them irrelevant.

File: nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/notification/http/HttpNotificationService.java (hunk 2 above)
Comment (nandorsoma) [reply nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/notification/http/HttpNotificationService.java:130]: Yes, it was not used.

File: nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/notification/http/HttpNotificationService.java (hunk 3 above)
Comment (nandorsoma) [reply nifi-bootstrap/src/main/java/org/apache/nifi/bootstrap/notification/http/HttpNotificationService.java:95]: Yes, I mentioned it in the summary.

**Comment by nandorsoma** [comment]
Hey @Lehel44!
Thank you for your review! I've answered your questions inline, which were related to functionality. About the Javadoc... I mainly used the original Javadoc and removed a part that was unrelated to this case. Because of that, I'd leave it as it is for now if it suits you.

**Review by Lehel44 — APPROVED** [review]
So as we agreed there's no need to change the javadoc now. Thanks for the changes. LGTM+1

**Comment by nandorsoma** [comment]
Thank you for your review @exceptionfactory! You are right, I didn't think about custom components. I've removed the new defaultValue method.

**Review by turcsanyip — APPROVED** [review]
@nandorsoma Thanks for the review changes!
@Lehel44 @exceptionfactory Thanks for your valuable review comments!

+1 LGTM
Merging to main.

**Comment by EndzeitBegins** [comment]
Hey @nandorsoma, just stumbled upon this PR and wanted to let you know, that #8102 introduced the `defaultValue` method you proposed here originally. 

It will be available in the upcoming major release version 2.x, as we can make the breaking change introduced by the method overload then.

**Comment by nandorsoma** [comment]
> Hey @nandorsoma, just stumbled upon this PR and wanted to let you know, that #8102 introduced the `defaultValue` method you proposed here originally. 
> 
> 
> 
> It will be available in the upcoming major release version 2.x, as we can make the breaking change introduced by the method overload then.

Thanks for letting me know! GJ!

**Pull request closed without merging**